File: gitdot-web/packages/gitdot-api/src/endpoint/repository/get-repository-resources.ts
import { z } from "zod";
import { RepositoryResourcesResource } from "../../resource";
import type { Endpoint } from "../endpoint";

export const GetRepositoryResourcesRequest = z.object({
  last_commit: z.string().optional(),
  last_updated: z.iso.datetime().optional(),
});
export type GetRepositoryResourcesRequest = z.infer<
  typeof GetRepositoryResourcesRequest
>;

export const GetRepositoryResourcesResponse = RepositoryResourcesResource;
export type GetRepositoryResourcesResponse = z.infer<
  typeof GetRepositoryResourcesResponse
>;

export const GetRepositoryResources = {
  path: "/repository/{owner}/{repo}/resources",
  method: "POST",
  request: GetRepositoryResourcesRequest,
  response: GetRepositoryResourcesResponse,
} as const satisfies Endpoint;
export type GetRepositoryResources = typeof GetRepositoryResources;

File: gitdot-api/src/endpoint/repository/get_repository_resources.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};

use crate::{endpoint::Endpoint, resource::repository::RepositoryResourcesResource};

pub struct GetRepositoryResources;

impl Endpoint for GetRepositoryResources {
    const PATH: &'static str = "/repository/{owner}/{repo}/resources";
    const METHOD: http::Method = http::Method::POST;

    type Request = GetRepositoryResourcesRequest;
    type Response = GetRepositoryResourcesResponse;
}

#[derive(ApiRequest, Debug, Serialize, Deserialize)]
pub struct GetRepositoryResourcesRequest {
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub last_commit: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub last_updated: Option<DateTime<Utc>>,
    #[serde(default)]
    pub force_refresh: bool,
}

pub type GetRepositoryResourcesResponse = RepositoryResourcesResource;
